### PR TITLE
Task 3 from networking ready for review

### PR DESCRIPTION
**Pull request by qnislavnachev**

### CODE REVIEW

File: tests/client/ClientTest.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+                client.join("localhost", 1111);
+            }
+        });
+        fakeServer.start();
+    }
+
+    @Test
+    public void ClientReceiveMessage() throws Exception {
+        clientThread.start();
+        Thread.sleep(1000);
```
Comment (tsony-tsonev) [inline tests/client/ClientTest.java]: you should never sleep the thread in the test, also I don't see test for sending messages from the client

File: tests/server/ServerTest.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        assertThat(actual, is(expected));
+        server.shutDown();
+    }
+
+    @Test
+    public void ServerSendMessageToAll() throws Exception {
+        fakeClient.start();
+        Thread.sleep(1000);
+        fakeClient1.start();
+        Thread.sleep(1000);
```
Comment (tsony-tsonev) [inline tests/server/ServerTest.java]: same here remove the sleeps, also add tests for receiving messages

File: src/task3/client/Client.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            @Override
+            public void run() {
+                try {
+                    BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+                    while ((message = reader.readLine()) != null) {
+                        System.out.println(message);
+                    }
+                } catch (SocketException ex) {
+                    System.out.println("Server fall down !");
+                    System.exit(0);
```
Comment (qnislavnachev) [inline src/task3/client/Client.java]: this System.exit(0) is for when I close server it should close the clients too, without it  the scanner in client still waiting for input

File: tests/client/ClientTest.java (hunk 1 above)
Comment (qnislavnachev) [inline tests/client/ClientTest.java]: can I use mock objects to remove this sleep ?

File: src/task3/client/Client.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import java.util.concurrent.BlockingQueue;
+
+public class Client extends Thread {
+    private String message;
+    private String host;
+    private int port;
+    private InputStream inputStream;
+    private BlockingQueue<String> queue;
+
+    public Client(InputStream inputStream, BlockingQueue<String> queue) {
```
Comment (tsony-tsonev) [inline src/task3/client/Client.java]: You should never add logic to the production code just to make your tests pass, remove it from here and from the server.

Comment (qnislavnachev) [inline src/task3/client/Client.java]: fixed

File: tests/client/ClientTest.java (hunk 1 above)
Comment (qnislavnachev) [inline tests/client/ClientTest.java]: fixed

File: src/task3/server/Server.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+    public void run() {
+        try {
+            server = new ServerSocket(port);
+            System.out.println("Server is set up... ");
+            while (true) {
+                Socket socket = server.accept();
+                listOfClients.add(socket);
+                System.out.println("New client detected");
+                Sender sender = new Sender(listOfClients);
```
Comment (qnislavnachev) [inline src/task3/server/Server.java]: fixed

File: src/task3/client/Client.java (hunk 3 above)
Comment (qnislavnachev) [inline src/task3/client/Client.java]: fixed

File: src/task3/client/Client.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    public String getMessage() {
+        return message;
+    }
+
+    public void join(String host, int port) {
+        this.host = host;
+        this.port = port;
+        start();
```
Comment (tsony-tsonev) [inline src/task3/client/Client.java]: why you need this

File: src/task3/server/Server.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+    public void run() {
+        try {
+            server = new ServerSocket(port);
+            System.out.println("Server is set up... ");
+            while (true) {
+                Socket socket = server.accept();
+                listOfClients.add(socket);
+                System.out.println("New client detected");
+                ClientNotifier notifier = new ClientNotifier(listOfClients);
```
Comment (tsony-tsonev) [inline src/task3/server/Server.java]: why ClientNotifier is extracted as separate class and the serverListener is not.

File: src/task3/server/Server.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+                notifier.start();
+                serverListener(socket);
```
Comment (tsony-tsonev) [inline src/task3/server/Server.java]: actually you listen for messaged in the thread of the server this way the server will block no more clients will be able to connect, both the notification for the clients and the listening for messages should be in separate thread

File: tests/client/ClientTest.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    @Test
+    public void ClientReceiveMessage() throws Exception {
+        startClient();
+        queue.take();
+        String actual = client.getMessage();
+        String expected = "Server message";
+        assertThat(actual, is(expected));
+        fakeServer.shutDown();
```
Comment (tsony-tsonev) [inline tests/client/ClientTest.java]: move the starting of the client and the server shutdown in the `@Before` and `@After` methods

**Comment by tsony-tsonev** [comment]
LGTM

**Pull request merged**